### DISCUSSION TITLE
ABNF for standard stack notation

### DISCUSSION THREAD

**Discussion by ruv** [ForthHub/discussion#171] [Show and tell]
## Stack diagrams

Stack diagrams are a kind of **documentation**. They allow us to shorten the text description and quickly understand the parameters of a word.

Stack diagrams describe input and output stack parameters for a Forth definition (or some semantics of a named Forth definition). We can think of the actual input stack parameters as a _tuple of parameters_, and the actual output stack parameters as a tuple of parameters _corresponding_ to the input tuple. A tuple of parameters can be an empty tuple.

Standard stack diagrams are written in special [**stack notation**](https://forth-standard.org/standard/notation#subsection.2.2.2). This stack notation is a formal language.

An example of a standard stack diagram: `search-wordlist ( c-addr u wid -- 0 | xt 1 | xt -1 )`.

A standard stack diagram describes _data types_ of input and output tuples. Formally, it describes a [function type](https://en.wikipedia.org/wiki/Function_type) using [union types](https://en.wikipedia.org/wiki/Union_type) and [product types](https://en.wikipedia.org/wiki/Product_type).

## Stack notation

The stack notation describes elements of a stack diagram, what they mean, and how they can be combined.

A stack diagram consists of data type symbols (with optional labels), operators, and markup characters.

An example of a stack diagram: `( S: foo -- bar )`. It specifies that a tuple of cells that has the type `foo` (as a whole) is consumed from the data stack, and a tuple of cells that has the type `bar` (as a whole) is produced on the data stack.  Typically, `foo` and `bar` are some expressions over _data type symbols_.

A **data type symbol** identifies a _data type_. Data type symbols are defined in [Table 3.1: Data types](https://forth-standard.org/standard/usage#table:datatypes) and in the corresponding section of each word set specification (if any). Examples: `x`, `xt`, `wid`, `n`, `r`.

A literal numeric value denotes a [singleton](https://en.wikipedia.org/wiki/Singleton_set), a [unit data type](https://en.wikipedia.org/wiki/Unit_type) that is a subtype of _n_ or _r_ (depending on the stack and/or literal form), and whose only member is this value. Examples: `0`, `1`, `-1`, `0e`, `1e`.

A concatenation of two data type symbols with white space characters between them represents a [product type](https://en.wikipedia.org/wiki/Product_type). So, a nonempty sequence of white space character denotes the binary product operator over data types. Examples:  `( xt n )`, `( c-addr u wid )`.

There is a short notation for a product type over the same data type — n-power type. When _n_ is a fixed number, this type is represented by concatenation of the number of power, the character «`*`» (star), and a data type symbol. For example: `( 3*x )`, which is equivalent to `( x x x )`. Note that  `( 0*x )` is equivalent to `( )`, and this denotes a [unit data type](https://en.wikipedia.org/wiki/Unit_type) whose only member is the empty data object, which occupies zero cells on a stack. It means, all the following expressions are equivalent: `( x 0*x 0*x )`, `( x 0*x )`, `( 0*x x 0*x )`, `( x )`. 

A concatenation of two data type symbols with a separator «`|`» (vertical bar) between them represents a _union type_. So, a vertical bar denotes the binary union operator over data types. Examples: `( n|u )` (in a short notation), `( 0 | xt 1 | xt -1 )` (in a long notation). Note that a missed data type symbol denotes the unit type `0*x`. For example all the following expressions are equivalent: `( x| )`, `( x|0*x )`, `( x |  )`, `( x | 0*x )`.

There is a notation for an n-power type when the power is not fixed. In this case a lowercase Latin letter like `i`, `j`, `k` is used instead of a number. For example `( i*x )`, `( k*r )`. Such an expression represents a union-type of n-power types from _n = 0_ to an implementation-defined limit (say, 32767). So, `( i*x )` is equivalent to `( 0*x | 1*x | 2*x | ... | 32767*x )`.

Note that `( 0*x 1*x )` is equivalent to `( 1*x )`, that is equivalent to `( x )`, but `( 0*x | 1*x )` **are not** equivalent to `( 1*x )`.

## Stack notation grammar

The stack notation grammar can be formally specified using ABNF (see <a href="https://www.rfc-editor.org/rfc/rfc5234">RFC5234</a>), as follows.

(Please note that there may be some mistakes in this specification)

```abnf
ALPHA-UPPER =  %x41-5A ; A-Z
ALPHA-LOWER =  %x61-7A ; a-z
ALPHA =  ALPHA-UPPER / ALPHA-LOWER ; A-Z / a-z
DIGIT =  %x30-39 ; 0-9
DQUOTE = %x22 ; `"`

quoted-string = DQUOTE *( %x20-21 / %x23-7E ) DQUOTE ; e.g. `"foo bar"`
whitespace = %x01-20 ; any allowed white space character
ws = 1*( whitespace ) ; one or more white space characters

stack-diagram = "(" ws  [ stack-id ":" ws ]  function-type ")"
stack-id = "S" / "R" / "C" / "F" ; an identifier of a stack
; There are the data stack, return stack, control-flow stack, floating-point stack.
; If a stack-id is absent, the data stack is implied.

function-type = left "--" ws right [ "|" ws function-type ] ;  e.g. ( +n -- true | n -- false )
left = product-type [ parsed-text-abbreviation ] ; a type of the input tuple and optional parsed text
right = union-type ; a type of the output tuple
parsed-text-abbreviation = quoted-string ws [ parsed-text-abbreviation ]

union-type = product-type [ "|" ws union-type ] ; the long notation for a union type, e.g.: `xt -1 | 0`
product-type = data-type-union ws [ product-type ] ; the long notation for a product type
; NB the long notation of a union type cannot be used for the input tuple,
; it can be only used for the output tuple.
; The variant ( foo -- bar | baz ) is equivalent to  (  foo -- bar  |  foo --  baz ).
; The variant ( foo | bar -- baz ) is incorrect. Either use ( foo -- baz | bar -- baz ),
; or use a short notation for a union type ( foo|bar -- baz )
; or introduce and employ a new data type symbol as ( foo_or_bar -- baz )
; where   foo_or_bar = ( foo | bar )

data-type-union = data-type-product [ "|" data-type-union ] ; the short notation for a union type, e.g.:  `u|n`, `ud|d`
data-type-product = [ tuple-length "*" ] data-type-element / "" ; the short notation for a product type, e.g. `i*x`, `2*n`
; NB: if a data-type-symbol is absent, the data type whose only member is the empty tuple is implied,
; for example: `x| `,  ( x -- x | )

tuple-length = tuple-length-variable / 1*DIGIT ; a variable name or non-negative integer number
tuple-length-variable = "i" / "j" / "k" / "l" / "m" / "n" ; may be other?

data-type-element = data-type-singleton / data-type-symbol [ element-label ]
; e.g.  `2`, `-1`, `n1`, `d.2`, `n.minuend`
data-type-singleton = [ "-" ] 1*DIGIT ; a signed integer number (a literal)
data-type-symbol = ( ALPHA-LOWER / "+" / "_" ) [ *( "-" ) data-type-symbol ] ; e.g. `+n`, `colon-sys`
; NB: no digits are allowed in a `data-type-symbol` to avoid confusing with an `element-label-numeric`
element-label =  element-separator element-label-any  /  element-label-numeric 
element-separator = "."
element-label-any = ( ALPHA-LOWER / DIGIT / "_" ) [ *( "-" / "." ) element-label-any ]
element-label-numeric = 1*DIGIT ; one or more digits
```

A `data-type-symbol` is a standard data type symbol or a user-defined data type symbol.

An absent `data-type-symbol` implies the data type `0*x` whose only member is the empty tuple of cells. It is a [unit](https://en.wikipedia.org/wiki/Unit_type) data type.

Note that an `element-label` can be rendered without a `separator` as a subscript.

## <a name="symbol-labels">Labels in data type symbols</a>

In a text description we need to refer one or another particular stack parameter.  There are few methods to do it.

A stack parameter can be referred to in a text description as follows:
1. by specifying the parameter kind (input or output) and position from the top of the stack (traditionally, starting from 1, i.e., the topmost parameter is the first parameter);
2. by literally specifying the corresponding element of the stack diagram.

We need **labels** along with data type symbols in a stack diagram **only** to *unambiguously* refer to the corresponding stack parameters in the text description when the same data type symbol is used in different elements of the stack diagram.

For example: `my-minus ( n -- n n )` — is a correct stack diagram. But we cannot use «_n_» to refer to a particular stack parameter.

If we add the labels in the example above: `my-minus ( n.1 n.2 -- n.3 )` — then each stack parameter can be uniquely referenced in the text description using the corresponding data type symbol and label from the stack diagram: _n.2_ refers to the first input parameter, _n.1_ refers to the second input parameter, _n.3_ refers to the first output parameter. NB: the value of a numeric label **says nothing about the position** in the stack of the corresponding stack parameter.

If a data type symbol is used only once in a stack diagram, or if we don't need to refer to the corresponding stack parameter in the text description, then we don't need to add a label to the data type symbol in the stack diagram. However, we can add a descriptive label to give additional clues what the parameter means.

An example of descriptive labels: `my-minus ( n.minuend n.subtrahend -- n.difference )`.

The same data type symbol in two or more elements of a stack diagram means that the corresponding stack parameters belong to this (the same) data type. And nothing more.  It does not mean that the parameters in the corresponding positions are equal.

The same data type symbol **and** the same label in two or more elements of a stack diagram means that the corresponding stack parameters are equal.

For example, in this case: `my-func ( n1 n2 -- n2 n3 )`,  we can use _n1_, _n2_, and _n3_ to refer to the corresponding stack parameters. Also, this diagram means that the stack parameters in the positions that correspond to the element _n2_ of the stack diagram are equal in both type and value. When _n2_ is used in a text description, it refers to one actual stack parameter value, that is the first input parameter and the second output parameter.

-----

Update: terminology has been corrected as noted [below](https://github.com/ForthHub/discussion/discussions/171#discussioncomment-11030626).

**Comment by ruv** [comment]
A stack notation grammar that allows multiple stacks in one stack diagram.

```abnf
ALPHA-UPPER =  %x41-5A ; A-Z
ALPHA-LOWER =  %x61-7A ; a-z
ALPHA =  ALPHA-UPPER / ALPHA-LOWER ; A-Z / a-z
DIGIT =  %x30-39 ; 0-9
DQUOTE = %x22 ; `"`

quoted-string = DQUOTE *( %x20-21 / %x23-7E ) DQUOTE ; e.g. `"foo bar"`
whitespace = %x01-20 ; any allowed white space character
ws = 1*( whitespace ) ; one or more white space characters

stack-diagram = "(" ws function-type-union ")"
function-type-union = function-type [ "|" ws function-type-union ] ;  e.g. ( +n -- true | n -- false )
function-type = [ stack ] function-type-single-stack *( stack function-type-single-stack )
stack = stack-id ":" ws
stack-id = "S" / "R" / "C" / "F" ; an identifier of a stack
; There are the data stack, return stack, control-flow stack, floating-point stack.
; If a stack-id is absent, the data stack is implied.

function-type-single-stack = input-type "--" ws output-type
input-type = product-type [ parsed-text-abbreviation ] ; a type of the input tuple and optional parsed text
output-type = union-type ; a type of the output tuple
parsed-text-abbreviation = quoted-string ws [ parsed-text-abbreviation ]

union-type = product-type [ "|" ws union-type ] ; the long notation for a union type, e.g.: `xt -1 | 0`
product-type = data-type-union ws [ product-type ] ; the long notation for a product type
; NB the long notation of a union type cannot be used for the input tuple,
; it can be only used for the output tuple.
; The variant ( foo -- bar | baz ) is equivalent to  (  foo -- bar  |  foo --  baz ).
; The variant ( foo | bar -- baz ) is incorrect. Either use ( foo -- baz | bar -- baz ),
; or use a short notation for a union type ( foo|bar -- baz )
; or introduce and employ a new data type symbol as ( foo_or_bar -- baz )
; where   foo_or_bar = ( foo | bar )

data-type-union = data-type-product [ "|" data-type-union ] ; the short notation for a union type, e.g.:  `u|n`, `ud|d`
data-type-product = [ tuple-length "*" ] data-type-element / "" ; the short notation for a product type, e.g. `i*x`, `2*n`
; NB: if a data-type-symbol is absent, the data type `0*x` whose only member is the empty tuple is implied,
; for example: `x| `,  ( x -- x | )

tuple-length = tuple-length-variable / 1*DIGIT ; a variable name or non-negative integer number
tuple-length-variable = "i" / "j" / "k" / "l" / "m" / "n" ; may be other?

data-type-element = data-type-singleton / data-type-symbol [ element-label ]
; e.g.  `2`, `-1`, `n1`, `d.2`, `n.minuend`
data-type-singleton = [ "-" ] 1*DIGIT ; a signed integer number (a literal)
data-type-symbol = ( ALPHA-LOWER / "+" / "_" ) [ *( "-" ) data-type-symbol ] ; e.g. `+n`, `colon-sys`
; NB: no digits are allowed in a `data-type-symbol` to avoid confusing with an `element-label-numeric`
element-label =  element-separator element-label-any  /  element-label-numeric 
element-separator = "."
element-label-any = ( ALPHA-LOWER / DIGIT / "_" ) [ *( "-" / "." ) element-label-any ]
element-label-numeric = 1*DIGIT ; one or more digits
```

Usage example:
```forth
: ?exit
  \ Compilation: ( colon-sys i*x -- colon-sys i*x )
  \ Run-time: ( S: 0 --  |  S: nz -- never  R: nest-sys -- )
  postpone if postpone exit postpone then
; immediate
```

**Rationale:** the single stack diagram `( S: 0 --  |  S: nz -- never  R: nest-sys -- )` specifies a more narrow type than two independent stack diagrams: `( 0 -- | nz -- never ) ( R: -- | nest-sys -- never )`.

-----

Update: terminology has been corrected as noted [below](https://github.com/ForthHub/discussion/discussions/171#discussioncomment-11030626).

**Reply by SirWumpus** [reply]
What of the case where the data-stack and float-stack is shared?  Is there one diagram for that case and separate one for split stacks?

**Reply by ruv** [reply]
If you need to specify the stack diagram for both cases: when the floating-point stack is shared with the data stack, and when it is separate, then a separate group of stack diagrams should be specified for each case.

Note that most words in [The optional Floating-Point word set](https://forth-standard.org/standard/float) have two groups of stack diagrams.
For example, for the [word `f!`](https://forth-standard.org/standard/float/FStore) the following diagrams are specified:
  - `( f-addr -- ) ( F: r -- )` — for the case of separate stacks;
  - `( r f-addr -- )` — for the case of a shared stack.

However, the first group could be specified as a single diagram to represent the single data type, instead of two data types (no changes in the second group):
  - `( S: f-addr --  F: r -- )`
  - `( r f-addr -- )`



NB: As of Forth-2012, the floating-point stack in a standard Forth system is separate, but a standard Forth system may have an environmental restriction on providing the shared floating-point stack.

**Reply by ruv** [reply]
For readability, it might make sense to separate different stacks in a single stack diagram with `;`.
For example,
- Instead of
  - `( S: f-addr --   F:  r -- )`
- write
  - `( S: f-addr -- ;  F:  r -- )`
  
Do you think the latter option is better than the former?

**Reply by SirWumpus** [reply]
I think the latter is clearer to read and I think would be better to parse.  In some cases terms in the previous list might be confused with the stack specifier.  

Sometimes I personally prefer to name stack items instead of use a type, eg. `(S: src s dst d --- bool )`  to distinguish between similar terms.  I suppose when naming terms, one could pre/post-fix the terms with a type too, eg. `(S: src:a s:u dst:a d:u -- bool:f )`, to be more clear.  

So when you give two stacks in the diagram `(S: src:a s:u dst:a d:u -- bool:f ; F: foo:f bar:f -- )` a list separator helps.

**Reply by ruv** [reply]
> I think the latter is clearer to read and I think would be better to parse. 

For parsing it does not matter.  Because non-space characters between `;` and a stack-id are not valid, and this should be checked.

>  In some cases terms in the previous list might be confused with the stack specifier.

Yes, that is an argument.

> I suppose when naming terms, one could pre/post-fix the terms with a type too, eg. `(S: src:a s:u dst:a d:u -- bool:f )`, to be more clear.

This is the idea of [labels](https://github.com/ForthHub/discussion/discussions/171#user-content-symbol-labels) along data type symbols. The labels are optional. In the text of the standard the labels are subscripts. In plain text labels can be separated by a dot (as example). Then, you example will become: `( addr.src u.src addr.dst u.dst -- flag )`

If we use the data type symbol `sd` for a character string cell pair `( c-addr u )`,  your example will become: `( sd.src sd.dst -- flag )`.

**Comment by ruv** [comment]
Currently, in the stack notation, the precedence among the operators «`--`», «`|`», «` `» (white space) is defined by the grammar. There are no any dedicated syntax to specify a sub-expression and override the order of operators (like parentheses in the mathematical infix notation).

Technically, a syntax for sub-expressions can be introduced, for example, using curly brackets. Then the expression `( xt 1 | xt -1 )` could be written as `( xt { 1 | -1 } )`.  I do not think such syntax is worth introducing, because it is very rarely needed. And from the readability point of view, it is better to introduce new data type symbols in complex cases (if any).

**Reply by ruv** [reply]
What can be useful in a stack diagram is a notation for the relative complement (which is similar to [that operation on sets](https://en.wikipedia.org/wiki/Complement_(set_theory))).  If we use the character «`\`» to denote the relative complement, then the expression `A\B` means the relative complement of _B_ with respect to _A_, and it represents a data type that includes all those and only those members of a data type _A_ that are not members of a data type _B_.

This allows us to specify data types in a stack diagram more narrowly. For example, instead of `( +n -- true | n -- false )` we can write `( +n -- true | n\+n -- false )` that means that _false_ is returned only for a negative input parameter.

Update: note that the data type symbol `+n` identifies **non-negative** integer numbers, see [3.1 Data types](https://forth-standard.org/standard/usage#usage:data).

**Reply by ruv** [reply]
### About precedence

- According to the [specified grammar](https://github.com/ForthHub/discussion/discussions/171#discussioncomment-10640300), the stack diagram:
  - `( foo -- bar | baz | fizz -- qux )`, where `foo`, `bar`, `baz`, `qux` are metasyntactic variables for some simple product types,
- shall be interpreted as:
  - `(   {  foo -- { bar | baz }  }   |   { fizz -- qux }   )`, where curly brackets are used for grouping.
  - This means that `fizz` is not a parameter of an application of the sum operator «`|`», but a parameter of an application of the function operator «`--`».

Such expressions, where the sum operator (in the long notation) is used both on product types and on function types in the same stack diagram, are rarely used in practice and can be very confusing to the reader. Therefore, it is better to avoid such expressions and use equivalent forms.
- An equivalent form for the example above is:
  - `(  foo -- bar  |  foo -- baz  |  fizz -- qux  )`

**Reply by ruv** [reply]
### More about precedence

Note that all the below rules are determined by [the grammar](https://github.com/ForthHub/discussion/discussions/171#discussioncomment-10640300).

There are two notations for the **sum operator**: short and long.
  - example of short notation: «`u|n`»  (no spaces);
  - example of long notation: «`u | n`» (elements are separated by spaces).

For the **power operator** (and unspecified power), only a short notation is used, e.g.: «`2*n`», «`i*x`» (no spaces).

For the **product operator**, only a long notation is used, e.g.: «`c-addr u wid`» (separated by spaces).

For the **function operator**, only a long notation is used, e.g.: «`foo -- bar`» (separated by spaces).

1\. A short notation always takes precedence over a long notation.
  - `( 2*bar | baz )` is interpreted as `( { bar bar } | baz )`;
  - `( 2*bar baz )` is interpreted as `( { bar bar } baz )`, which is equivalent to `( bar bar baz )`;
  - `( foo|bar baz )` is interpreted as `( { foo | bar } baz )` that is equivalent to `( { foo baz }  |  { bar baz } )`;
  - `( foo|bar | baz )` is interpreted as `( { foo | bar } | baz )`, which is equivalent to `( foo | baz | baz )`;
  - `( foo|bar -- baz )` is interpreted as `( { foo | bar } -- baz )` that is equivalent to `( { foo -- baz } | { bar -- baz } )`.

2\. The product operator always takes precedence over the long notation of the sum operator.
  - e.g. `( foo | bar baz | qux )` is interpreted as `( foo | { bar baz } | qux )`.

3\. The product operator always takes precedence over the function operator.
  - e.g. `( foo bar -- baz qux )` is interpreted as `( { foo bar }  --  { baz qux } )`.

4\. The precedence between the function operator and the long notation of the sum operator depends on the context.
  - `( foo -- bar | baz )` is interpreted as `( foo -- { bar | baz } )`,
  - `( foo -- bar | baz -- qux )` is interpreted as `( { foo -- bar }  |  { baz -- qux } )`.

**Comment by alextangent** [comment]
What would zero return? I’d suggest false based on your description and the stack diagram, since zero is not a member of the set of positive integers. 

 

__ 

Alex

**Reply by ruv** [reply]
@alextangent wrote:
> What would zero return?

The data type _+n_ includes zero, despite its name. See [3.1 Data types](https://forth-standard.org/standard/usage#usage:data):
- `+n`		non-negative number

So, `( +n -- true | n\+n -- false )` means that when the input parameter is zero, the output parameter is true.

**Comment by ruv** [comment]
An excerpt from the [first post](https://github.com/ForthHub/discussion/discussions/171#discussion-7156477) (updated):
> Note that `( 0*x )` is equivalent to `( )`, and this denotes a [unit data type](https://en.wikipedia.org/wiki/Unit_type) whose only member is _the empty data object_, which occupies zero cells on a stack. It means, all the following expressions are equivalent: `( x 0*x 0*x )`, `( x 0*x )`, `( 0*x x 0*x )`, `( x )`.

It is important to distinguish between _the empty **data object**_ and _the empty **data type**_.

## Empty data object

Forth definitions consume and produce actual data objects on the stack during execution. Several data objects on the stack constitute a tuple of data objects.

**The empty data object** is a pseudo data object that is a tuple of zero bits. So, the empty data objects occupies zero stack cells.

If a Forth definition consumes nothing from the stack, it _formally_ consumes the empty data object. If a Forth definition produces nothing on the stack, it _formally_ produces the empty data object.

There is a data type whose only member is the empty data object. This data type can be denoted as `0*x`. In a stack diagram this data type may be omitted.   For example, the diagram `( -- )` means that nothing consumed, and nothing produced on the stack. An equivalent diagram is `( 0*x -- 0*x )`. 

Obviously, any tuple of empty data objects is equivalent to the empty data object. Hence, any product of `0*x` types is equivalent to `0*x` type. For example,  `( 0*x 0*x 0*x )` is equivalent to `( 0*x )`.


## Empty data type

**The empty data type** is a data type that **has no members** (it is [_uninhabited_](https://en.wikipedia.org/wiki/Empty_type)).  Thus, this data type is a sub-type of any data type (this follows from the definition of the _subtype relation_, and this is similar to how [the empty set](https://en.wikipedia.org/wiki/Empty_set#Properties) is a subset of any set). Hence, this empty type is a [_bottom type_](https://en.wikipedia.org/wiki/Bottom_type), which follows from the definition of "bottom type".


If we specify the empty data type as a type of an output parameter in a stack diagram for a definition, it effectively means that the definition does not return control to the caller. Because if it returns control, it returns some data object (possibly the empty data object).  But it is impossible to return a data object that is a member of the empty data type.  Then, it does not return control at all.

Similar, if we specify the empty data type as a type of an input parameter, it means that the definition never gets control. This is impossible for actual definitions, but it is possible for unreachable code fragments.

It is therefore useful to introduce a data type symbol for the empty data type, as in some cases it allows us to specify narrower stack diagrams or data types of smaller size.

Let `never` be a data type symbol for the empty data type. Then, the following stack diagrams can be specified for some standard words:

- `bye ( -- never )`
- `throw ( 0 -- | x\0 -- never )`
- `abort ( -- never )`
- `quit ( -- never )`
- `exit` Run-time `( -- never )`
- `again` Run-time `( -- never )`
- `while` Run-time `( 0 -- never | x\0 -- )`
- `until` Run-time `( 0 -- never | x\0 -- )`
- `if` Run-time `( 0 -- never | x\0 -- )`
- `abort"` Run-time `( 0 -- | x\0 -- never )`

Output of run-time semantics shows the input for a word that is compiled after these semantics (if any).

An example of a user-defined word:
```forth
: ?exit \ Run-time: ( 0 --  |  x\0 -- never )
  postpone if postpone exit postpone then
; immediate
```

**Comment by ruv** [comment]
## Changes in terminology

I looked deeper into the type theory and realized that what I called a "sum type" in this thread should be called a "union type".

A union type implies an untagged union, but a sum type usually implies a tagged union (aka disjoint union). Sometimes an untagged union is called a sum type, but this is confusing.

The difference is as follows.

Let the sum type operator be denoted by the symbol `+`, the union type operator be denoted by `|`. And let we have types: `TS = A + B`, and `TU = A | B`  (where `A` and `B` are some types).

If a function expects an argument of type `TS` (i.e., `A + B`), you **cannot** submit a parameter of type `A`, or a parameter of type `B`. You have to submit a parameter of type `A+B`.  **Because neither members of `A` nor members of `B` are members of `A + B`**. 
 
The consequence of this is that  `( A + A ) <> A`, and the order of the arguments of the sum-operator **matters**.

If a function expects an argument of type `TU` (i.e., `A | B`), you **can** submit a parameter of type `A`, or a parameter of type `B`. Because every member of `A` is a member of `A | B`, and every member of `B` is a member of `A | B`.

The consequence of this is that `( A | A ) = A`, and the order of the arguments of the union-operator **does not matter**.


### Discrimination of members

When using a parameter of union type `A | B`, we usually need some method to discriminate (distinguish) whether the actual parameter is of type `A` or type `B`. This is typically done with some custom tagging.  For example,  `( 0 | xt n\0 )` — the top cell of the tuple plays the role of a tag: when it is 0, this tuple consists of only one cell, when it is not 0, the length of the tuple is 2 and the second element is an _xt_.




### References

1. [Union Types in TypeScript](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#defining-a-union-type)
2. [Union Types in Python 3](https://docs.python.org/3/library/typing.html#typing.Union)
3. [Polymorphic union refinement](https://counterexamples.org/polymorphic-union-refinement.html#polymorphic-union-refinement)
4. [Formal definition of Sum type](https://ncatlab.org/nlab/show/sum+type) in type theory
